core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+            'LearnerActionRenderService',
+            'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+            'MAX_UNRELATED_ACTIONS_PER_BLOCK',
+            function(
+                $scope, $uibModalInstance, AlertsService,
+                LearnerActionRenderService,
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+                MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+              $scope.playthroughIndex = playthroughIndex;
+
+              // This is the index of the first action in the range of actions
+              // displayed. When this index reaches 0, we've displayed all the
+              // learner actions.
+              let indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return indexOfFirstDisplayedAction !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction ===
+                    playthrough.actions.length) {
+                  previousStateName = (
+                    playthrough.actions[
+                      0].actionCustomizationArgs.state_name.value);
+                } else {
+                  let action = playthrough.actions[indexOfFirstDisplayedAction];
+                  previousStateName =
+                    action.actionCustomizationArgs.state_name.value;
+                }
+                for (i = indexOfFirstDisplayedAction; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+
+                  if (currentStateName !== previousStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName
```

Comment:
Missing semicolon. Are the pre-submit linter checks running on your machine? /cc @oppia/dev-workflow-team

Reply:
@seanlip Nope, yet another import error. Will reach out on Gitter about that.

Reply:
@seanlip CircleCI did report the lint error.

Reply:
Yep, I know. But I thought this was supposed to run as a pre-push check on @pranavsid98's local machine ...?

(Pranav, not sure what you mean by "yet another import error". Could you please clarify?)

Reply:
Got this fixed. Thanks for pointing out the wiki